src/libdrakvuf/vmi.c

```diff
@@ -626,6 +626,33 @@ event_response_t int3_cb(vmi_instance_t vmi, vmi_event_t* event)
             loop = loop->next;
         }
     }
+
+    // Iterate over traps updating ttl.
+    GSList* update_ttl_loop = s->traps;
+    time_t cur_time = time(NULL);
+    while (update_ttl_loop)
+    {
+        drakvuf_trap_t* trap = (drakvuf_trap_t*) update_ttl_loop->data;
+        if (trap->ttl == UNLIMITED_TTL)
+        {
+            update_ttl_loop = update_ttl_loop->next;
+            continue;
+        }
+
+        if (cur_time - trap->last_ttl_rst >= TRAP_TTL_RESET_INTERVAL_SEC)
+        {
+            trap->last_ttl_rst = cur_time;
+            trap->ttl = drakvuf_get_limited_traps_ttl(drakvuf);
+        }
+
+        if (--trap->ttl == 0)
+        {
+            // Does not break the loop, as we just schedule traps to get
+            // removed later (in_callback == 1).
+            drakvuf_remove_trap(drakvuf, trap, NULL);
```

Comment:
You'll want to give an indicator to the plugins subscribing to this trap that the trap is being removed by hammering protection. Otherwise, how would the plugin know if it wants to re-enable the trap later? Can it even re-enable it?

Reply:
This is actually a really good point. We have discussed internally how to solve this, but couldn't find any clean solution that wouldn't involve many changes. Do you have any ideas in mind, other than adding another callback to trap structure?

Reply:
I was thinking perhaps the best would be if the plugin could specify how long a trap should be lifted when this trips. Then libdrakvuf could handle placing the trap back automatically.

Reply:
I've added an `API hammering callback` that will be invoked anytime a `TTL` will drop to 0. The default action is just to unhook the trap (as in most cases this is what we want to do). However, I leave it up to the plugin what action they want to perform – just log, or re-enable the hook.
My other idea was to not inform the plugin at all, just log such unhooking and reinsert the trap on libdrakvuf side after some period of time. The benefit of such an approach is that the user would not have to set these `ah_cb` field every time.
This would require adding some reference counting to the trap structure to prevent from deallocating the trap by plugin before it gets reinserted.